File: shared/python/src/dronefleet_shared/models/order.py
from datetime import datetime

from pydantic import BaseModel, Field

from .protocol import OrderPriority, OrderStatus
from .telemetry import Position


class Order(BaseModel):
    id: str
    pickup_location: Position
    delivery_location: Position
    status: OrderStatus = OrderStatus.PENDING
    priority: OrderPriority = OrderPriority.STANDARD
    product_type: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    assigned_drone_id: str | None = None
    assigned_mission_id: str | None = None
    solving_session_id: str | None = None

    @property
    def max_delivery_time_minutes(self) -> int:
        """Calculate deadline based on priority (minutes from creation)."""
        if self.priority == OrderPriority.CRITICAL:
            return 15
        elif self.priority == OrderPriority.HIGH:
            return 30
        return 60

    class Config:
        json_schema_extra = {
            "example": {
                "id": "ORDER-01",
                "pickup_location": {"lat": 48.85, "lon": 2.35},
                "delivery_location": {"lat": 48.90, "lon": 2.40},
                "product_type": "BLOOD",
                "priority": "HIGH",
            }
        }

File: shared/java/src/main/java/com/dronefleet/shared/models/Order.java
package com.dronefleet.shared.models;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Order {
    private String id;
    private Position pickupLocation;
    private Position deliveryLocation;
    private OrderStatus status;
    private OrderPriority priority;
    private String productType;
    private Instant createdAt;
    private String assignedDroneId;
    private String assignedMissionId;
    private String solvingSessionId;

    /**
     * Calculate delivery deadline based on priority.
     */
    public Instant getDeliveryDeadline() {
        if (createdAt == null) return Instant.now().plus(60, ChronoUnit.MINUTES);
        return switch (priority) {
            case CRITICAL -> createdAt.plus(15, ChronoUnit.MINUTES);
            case HIGH -> createdAt.plus(30, ChronoUnit.MINUTES);
            case STANDARD -> createdAt.plus(60, ChronoUnit.MINUTES);
            default -> createdAt.plus(60, ChronoUnit.MINUTES);
        };
    }

    /**
     * Calculate if order is overdue.
     */
    public boolean isOverdue(Instant now) {
        return now.isAfter(getDeliveryDeadline());
    }
}
